### PR TITLE
VPN-5472: Merge Mozilla.Shared.qmlcomponents into Mozilla.VPN module

### PR DESCRIPTION
**Pull request by oskirby**
## Description
We have received a few reports of users encountering a blank UI screen on some Linux distributions. So far this seems to manifest on Ubuntu 20.04 intermittently. When this occurs, we find a QML error in the log of the form: 
```
Warning: qrc:/ui/screens/ScreenHome.qml:9:1: module "Mozilla.VPN" is not installed (ScreenHome.qml:9)
```

However, this `Mozilla.VPN` QML module is created programmatically by the program. I suspect that this error occurs because something in the QML cache isn't able to figure out that this module exists, creating this module with `qt6_add_qml_module()` might do the trick.

## Reference
Github issue: #7912 ([VPN-5472](https://mozilla-hub.atlassian.net/browse/VPN-5472))

## Checklist
    
- [ ] My code follows the style guidelines for this project
- [ ] I have not added any packages that contain high risk or unknown licenses (GPL,  LGPL, MPL, etc. consult with DevOps if in question)
- [ ] I have performed a self review of my own code
- [ ] I have commented my code PARTICULARLY in hard to understand areas
- [ ] I have added thorough tests where needed

[VPN-5472]: https://mozilla-hub.atlassian.net/browse/VPN-5472?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by oskirby** [comment]
Bah, humbug. Okay, so this did actually fix the QML cache loading jankiness, but it completely broke the QML_NAMED_ELEMENT() classes from the C++ world. I guess this means I'm on the right track at least.

File: src/productshandler.cpp [hunk 1]
```diff
@@ -44,6 +44,11 @@ ProductsHandler::ProductsHandler(QObject* parent) : QAbstractListModel(parent) {
   Q_ASSERT(!s_instance);
   s_instance = this;
 
+  // If web purchases are supported, then we don't need to do anything.
```
Comment (mcleinman) [inline src/productshandler.cpp:47]: Why move this in?

File: nebula/ui/components/filterproxymodel.cpp [hunk 2]
```diff
@@ -4,12 +4,9 @@
 
 #include "filterproxymodel.h"
 
-#include "logger.h"
-#include "qmlengineholder.h"
-
-namespace {
-Logger logger("FilterProxyModel");
-}
+#include <QDebug>
```
Comment (mcleinman) [inline nebula/ui/components/filterproxymodel.cpp:7]: Why the changes in this file?

**Review by mcleinman — COMMENTED** [review]
Looks good. A couple questions so I can better understand this, as I don't know this part of the code super-well.

I don't have a Linux box or VM currently, so I'm not able to test it.

File: nebula/ui/components/filterproxymodel.cpp (hunk 2 above)
Comment (oskirby) [reply nebula/ui/components/filterproxymodel.cpp:7]: In order to move this class into the Nebula class, it was important to make sure that the C++ code doesn't depend on code in the `mozillavpn` target, otherwise this would result in a dependency loop. This mostly just required that we drop the use of the VPN's logging infrastructure and just do it in pure Qt.

File: src/addons/addonmessage.h [hunk 3]
```diff
@@ -22,8 +22,6 @@ constexpr const char* ADDON_MESSAGE_SETTINGS_STATUS_KEY = "state";
 class AddonMessage final : public Addon {
   Q_OBJECT
   Q_DISABLE_COPY_MOVE(AddonMessage)
-  QML_NAMED_ELEMENT(MZAddonMessage)
```
[the remark this answers is not in the document]
Comment (oskirby) [reply src/addons/addonmessage.h:25]: Generally I was finding that the QML types registered implicitly by the `mozillavpn` target were failing to make it into the QML namespace. To work around it, this PR is attempting to move all the QML type registration into the `mozillavpn-ui` target and we do it in `src/ui/types/MZAddonMessage.h` with the `QML_FOREIGN()` macro.

File: src/productshandler.cpp (hunk 1 above)
Comment (oskirby) [reply src/productshandler.cpp:47]: I couldn't find a way to conditionally register a singleton type purely using `qt_add_qml_module()` in a way that would respect the feature flags so I figured the cleanest solution was to register all the QML types in the `Mozilla.VPN` module (and to do so consistently regardless of feature flagging).

In order to replicate the intended behaviour of the C++ code that this replaces, it was necessary to ensure that the `ProductsHandler` class doesn't do anything when the feature flag is disabled even if we happen to instantiate it.

**Review by mcleinman — APPROVED** [review]
Thanks for those explanations. Code looks great to me, but I'm not able to test it - I don't have a Linux box set up. 

Naomi, your call on whether you want to merge or wait for additional 👍🏻 on this one.

File: src/commands/commandui.cpp [hunk 4]
```diff
@@ -309,61 +309,6 @@ int CommandUI::run(QStringList& tokens) {
...
-#endif
-
-    qmlRegisterSingletonInstance("Mozilla.VPN", 1, 0, "VPNPurchase",
-                                 PurchaseHandler::instance());
-
-    if (!Feature::get(Feature::Feature_webPurchase)->isSupported()) {
-      qmlRegisterSingletonInstance("Mozilla.VPN", 1, 0, "VPNProducts",
-                                   ProductsHandler::instance());
-    }
-
```
Comment (lesleyjanenorton) [inline src/commands/commandui.cpp:366]: This wasn't introduced by your changes but does it still make sense to register MZAccessibleNotification and MZI18n here instead of qmlengineholder? Seems odd to be handling this in two places.

File: nebula/ui/components/forms/MZSearchBar.qml [hunk 5]
```diff
@@ -6,7 +6,7 @@ import QtQuick 2.0
 import QtQuick.Layouts 1.14
 
 import Mozilla.Shared 1.0
-import Mozilla.Shared.qmlcomponents 1.0
+import Mozilla.VPN 1.0
```
Comment (lesleyjanenorton) [inline nebula/ui/components/forms/MZSearchBar.qml:9]: Do we still need the Mozilla.VPN import here

File: nebula/ui/components/MZGuideList.qml [hunk 6]
```diff
@@ -6,7 +6,7 @@ import QtQuick 2.15
 import QtQuick.Layouts 1.15
 
 import Mozilla.Shared 1.0
-import Mozilla.Shared.qmlcomponents 1.0
+import Mozilla.VPN 1.0
```
Comment (lesleyjanenorton) [inline nebula/ui/components/MZGuideList.qml:9]: Same question here. Do we still need the .VPN import?

**Review by lesleyjanenorton — COMMENTED** [review]
And here I was really hoping this was just an issue of adding an import statement or two. Oof. Thanks for taking this on @oskirby! 

A few things:
- Why move the composer dir into src/ui? Is it possible to move it to nebula/ui instead? NBD if not but so far only QML types in Nebula are prefixed with "MZ". (Another matter for our tech debt pile: we no longer need to separate the QML according to what can be 'shared' or not) 
- I'm still completely baffled as to why Mozilla.Shared works and Mozilla.VPN randomly fails (although to your point on Slack, maybe Shared fails too and we just don't know it yet).

**Comment by lesleyjanenorton** [comment]
Also want to note: My questions don't need to be answered if you're ready to merge. I'm just baffled and curious/wondering.

**Review by lesleyjanenorton — APPROVED** [review]

File: tests/qml/CMakeLists.txt [hunk 7]
```diff
@@ -161,8 +147,6 @@ target_sources(qml_tests PRIVATE
     ${MZ_SOURCE_DIR}/errorhandler.h
     ${MZ_SOURCE_DIR}/feature/feature.cpp
     ${MZ_SOURCE_DIR}/feature/feature.h
-    ${MZ_SOURCE_DIR}/filterproxymodel.cpp
```
[the remark this answers is not in the document]
Comment (oskirby) [reply tests/qml/CMakeLists.txt:164]: We do, it's just included as a part of the nebula library now instead of needing to be manually included in the tests.

File: src/commands/commandui.cpp (hunk 4 above)
Comment (oskirby) [reply src/commands/commandui.cpp:366]: I agree, it is a bit weird registering things in multiple places like this. However, my ideal solution, if I were up to another refactoring challenge, would be to combine the `Mozilla.VPN` and `Mozilla.Shared` modules together. They are vestigial remains of our attempt to modularize the codebase.

File: nebula/ui/components/qmldir [hunk 8]
```diff
@@ -15,8 +15,8 @@ MZCheckBox 0.1 MZCheckBox.qml
 MZCheckBoxRow 0.1 MZCheckBoxRow.qml
 MZChevron 0.1 MZChevron.qml
 MZClickableRow 0.1 MZClickableRow.qml
-MZCollapsibleCard 0.1 MZCollapsibleCard.qml
 MZComposerView 0.1 MZComposerView.qml
+MZCollapsibleCard 0.1 MZCollapsibleCard.qml
```
[the remark this answers is not in the document]
Comment (oskirby) [reply nebula/ui/components/qmldir:19]: Oops! I was messing around with moving the `MZComposerView` around a bunch, and I guess I forgot to put it back where I found it.

**Comment by oskirby** [comment]
>     * Why move the composer dir into src/ui? Is it possible to move it to nebula/ui instead? NBD if not but so far only QML types in Nebula are prefixed with "MZ". (Another matter for our tech debt pile: we no longer need to separate the QML according to what can be 'shared' or not)
I wanted to; it would have cleaned things up nicely and left all the composer stuff in the `nebula` library. Unfortunately, there's a dependency between the composer classes and the `AddonApi` which would result in a dependency loop between `nebula` and `mozillavpn`.
